framework/Optimizers/GradientDescent.py

```diff
@@ -0,0 +1,840 @@
+      @ Out, converged, bool, convergence state
+    """
+    stepSize = self._stepHistory[traj][-1]['magnitude']
+    converged = stepSize < self._convergenceCriteria['stepSize']
+    self.raiseADebug(self.convFormat.format(name='stepSize',
+                                            conv=str(converged),
+                                            got=stepSize,
+                                            req=self._convergenceCriteria['stepSize']))
+    return converged
+
+  def _checkConvObjective(self, traj):
+    """
+      Checks the change in objective for convergence
+      @ In, traj, int, trajectory identifier
+      @ Out, converged, bool, convergence state
+    """
+    if len(self._optPointHistory[traj]) < 2:
+      return False
+    o1, _ = self._optPointHistory[traj][-1]
+    o2, _ = self._optPointHistory[traj][-2]
+    delta = mathUtils.relativeDiff(o2[self._objectiveVar], o1[self._objectiveVar])
+    converged = abs(delta) < self._convergenceCriteria['objective']
+    self.raiseADebug(self.convFormat.format(name='objective',
+                                            conv=str(converged),
+                                            got=delta,
+                                            req=self._convergenceCriteria['objective']))
+    return converged
+  # END convergence Checks
+  # * * * * * * * * * * * * * * * *
+
+  def needDenormalized(self):
+    """
+      Determines if the currently used algorithms should be normalizing the input space or not
+      @ In, None
+      @ Out, needDenormalized, bool, True if normalizing should NOT be performed
+    """
+    return self._stepInstance.needDenormalized() or self._gradientInstance.needDenormalized()
+
+  def denormalizeGradient(self, gradMag):
+    """
+      Denormalizes the gradient to correspond to the original space.
+      @ In, gradMag, float, normalized space gradient magnitude
+      @ Out, denormed, float, original (denormalized) gradient magnitude
+    """
+    # if no normalization is occuring, then just return as is
+    if self.needDenormalized():
+      return gradMag
+    # scale by the product of the dimensions
+    scale = 1
+    for var in self.toBeSampled:
+      lower, upper = self._variableBounds[var]
+      scale *= upper - lower
+    return gradMag / scale
+
+  def _formatSolutionExportVariableNames(self, acceptable):
+    """
+      Does magic formatting for variables, based on this class's needs.
+      Extend in inheritors as needed.
+      @ In, acceptable, set, set of acceptable entries for solution export for this entity
+      @ Out, new, set, modified set of acceptable variables with all formatting complete
+    """
+    # remaking the list is easier than using the existing one
+    acceptable = RavenSampled._formatSolutionExportVariableNames(self, acceptable)
+    new = []
+    print('DEBUGG acceptable:', acceptable)
+    while acceptable:
+      template = acceptable.pop()
+      if '{CONV}' in template:
```

Comment:
are these special keywords explained anywhere? 
I am sure they are but I can not find them (the PR is quite big :D )

Reply:
They show up in the user manual documentation with descriptions, and they're self-fulfilled in each class in the formatting method and class method.